### PR TITLE
Redesign the `log` implementations

### PR DESCRIPTION
**Pull request by Thirumalai-Shaktivel**
Towards: https://github.com/lfortran/lfortran/issues/3034

### CODE REVIEW

File: tests/reference/asr-external4-4e8bc7b.stdout [hunk 1]
```diff
@@ -24,14 +24,20 @@
...
+                                    Local
+                                    ()
+                                    ()
+                                    Default
+                                    (Real 8)
+                                    ()
+                                    Source
+                                    Public
+                                    Required
+                                    .false.
```
Comment (Thirumalai-Shaktivel) [inline tests/reference/asr-external4-4e8bc7b.stdout:40]: @Pranavchiku Is this updated reference correct?

Comment (Pranavchiku) [reply tests/reference/asr-external4-4e8bc7b.stdout:40]: Yes looks good to me.

**Review by Pranavchiku — APPROVED** [review]
This is good. Thanks for doing this!

**Comment by Thirumalai-Shaktivel** [comment]
Cool, thanks for the review!
I will update the tests and merge.

**Comment by Thirumalai-Shaktivel** [comment]
Ready!

**Review by certik — APPROVED** [review]

**Comment by Thirumalai-Shaktivel** [comment]
Thanks for the review!

**Review by certik — APPROVED** [review]

**Pull request merged**